---
title: "Longitudinal Analysis"
author: "Kennedy Zinn"
date: "2025-07-21"
output: html_document
---

```{r}
packages <- c("minfi", "limma", "DMRcate", "maxprobes", "readxl", "dplyr", "missMethyl", "MethylToSNP", "glue", "IlluminaHumanMethylationEPICanno.ilm10b5.hg38", "msigdbr", "tibble", "tidyr", "clusterProfiler", "org.Hs.eg.db", "purrr", "GenomicRanges", "ggplot2")
sapply(packages, library, character.only = TRUE)
```

```{r}
cell_phase = "iPSC" # either iPSC or EC
```

# Load data
```{r}
# load sample sheet
sample_sheet <- read_excel("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/TS/dna methyl/sample_sheet_ALL.xlsx",
                         skip = 9,
                         na="NA"
                         )
sample_sheet <- sample_sheet[,1:10] %>% 
  dplyr::filter(Phase==cell_phase) %>% 
  dplyr::filter(Karyotype!="46XisoXq") %>% 
  dplyr::filter(Phase!="NA")
sample_sheet$Basename <- paste(sample_sheet$Sentrix_ID, sample_sheet$Sentrix_Position, sep = "_")
rownames(sample_sheet) <- sample_sheet$Basename
```

```{r}
# load the idat files
path_to_idat <- "/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/TS/dna methyl/SPMTS2_Array data/"

rgSet <- read.metharray.exp(path_to_idat, targets = sample_sheet)
```

# Pre-processing

```{r}
source("process_methyl_microarray.R")
gmSet <- process_methyl_microarray(rgSet)
```

Beta values, which range between [0,1], represent the proportion of methylated DNA copies at a specific CpG site compared to all of the DNA copies at that site. Beta values for the ith CpG site are mathematically defined as,

Beta(i) = max(yi(methyl),0) / [max(yi(unmethyl), 0) + max(yi(methyl), 0) + alpha],

where yi methyl and unmethyl are the intensities measured by the ith methylated and unmethylated probes, respectively. 0 ensures that negative signals that take place after background adjustment are not considered. Alpha, a constant (usually 100), serves as an offset to regularize beta when both methylated and unmethylated instensities are low. Beta values are intuitive, and hence useful for visual presentations of the reads. (Du et al., 2010)

```{r, echo = FALSE}
# Compute beta values
beta_values <- getBeta(gmSet) 
beta_values <- beta_values[,colnames(beta_values) %in% rownames(sample_sheet)]
```

M values are log transformed beta values. Negative M values indicate higher levels of unmethylated DNA copies than methylated DNA copies. Postive M values indicate higher levels of methylated DNA copies than unmethylated. M values close to zero indicate similar intensities between methylated and unmethylated probes. M values are mathematically expressed as,

M(i) = log2([max(yi(methyl))+alpha]/[max(yi(unmethyl))+alpha])

where alpha is a constant to offset dramatic changes resulting from small estimation errors. M values can be expressed in terms of beta values,

M(i) = log2(Beta(i)/[1 - Beta(i)])

and therefore, beta values can be expressed in terms of M values,

Beta(i) = 2\^M(i) / [2\^M(i) + 1]

M values are more statistically sound than beta values.

```{r}
# Compute M values
Mval <- getM(gmSet)
Mval <- Mval[,colnames(Mval) %in% rownames(sample_sheet)]
dim(Mval)
```


```{r}
# PCA of M values
disease <- factor(sample_sheet$Karyotype, levels = c("46XX", "45X"))

pdf(glue("MDS_plot_{cell_phase}.pdf"))
par(mfrow=c(1,1), cex =.5)
plotMDS(Mval, labels=sample_sheet$Sample_Name, col=as.numeric(disease))
legend("top",legend=c("46XX", "45X"), col =1:2)
dev.off()
```

# Differential Methylation Analysis

```{r}

# factor covariates
if(cell_phase == "iPSC"){
  patient <- factor(ifelse(sample_sheet$Patient=="TS050", "X50.53", "X14"))
} else {
  patient <- factor(ifelse(sample_sheet$Patient=="TS019", "X19", "X02.50"))
}
```

```{r}
design <- model.matrix(~disease+patient)
design
```

```{r}
fit <- lmFit(Mval, design)
fit2 <- eBayes(fit)
top <- topTable(fit2, number = "Inf", coef="disease45X") # diff in karyotype, while controlling for patient
top
```

```{r}
# print proportion of sig CpGs hyper and hypo methylated
sig.cpg <- top[top$adj.P.Val<0.05,]
t <-table(sig.cpg$logFC<0)

hyper.proportion <- t[1]/nrow(sig.cpg)
hypo.proportion <- t[2]/nrow(sig.cpg)

glue("{nrow(sig.cpg)} probes are significantly differentially methylated in 45XO cells compared to 46XX cells. The proportion of hypermethylated CpG sites is {hyper.proportion}. The proportion of hypomethylated CpG sites is {hypo.proportion}.")
```

## Kegg Region Analysis
```{r}
myAnnotation <- cpg.annotate(datatype = "array", object = Mval, what = "M",
                             arraytype = "EPICv1",
                             analysis.type = "differential", design = design, contrasts = F, coef = 2)
DMRs <- suppressMessages(dmrcate(myAnnotation))
results.ranges <- extractRanges(DMRs, genome="hg38")
```

```{r}
anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)

gst.region.kegg <- goregion(results.ranges, all.cpg = rownames(top), collection = "KEGG", array.type="EPIC", anno=anno, sig.genes=T)
gst.region.go <- goregion(results.ranges, all.cpg = rownames(top), collection = "GO", array.type="EPIC", anno=anno, sig.genes=T)

top.go <- topGSA(gst.region.go, number = "Inf")
top.go <- top.go[top.go$FDR<0.05,]

top.kegg <- topGSA(gst.region.kegg, number = "Inf")
top.kegg <- top.kegg[top.kegg$FDR<0.05,]
```

```{r}
# Get the Hallmark gene sets for Homo sapiens (Human)
m_df <- msigdbr(species = "Homo sapiens", category = "H")

# Filter for the specific gene set and get the gene list
emt_gene_list <- m_df %>%
  filter(gs_name == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>%
  pull(entrez_gene)

# Put the gene list into a named list, as required by gsaregion
hallmark_gs_list <- list("HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION" = emt_gene_list)
```

```{r}
gsa.region <- gsaregion(results.ranges, all.cpg = rownames(top), collection = hallmark_gs_list, array.type="EPIC", anno=anno, sig.genes=T)
```

# Pathway Analysis by Methylation Direction

## Identify the direction of significant gene sets
```{r}
sig.Genes.all <- unique(c(gst.region.go$SigGenesInSet,
                         gst.region.kegg$SigGenesInSet))
sig.genes.endmt <- unique(gsa.region$SigGenesInSet)

sig.gene.list <- function(sig.GeneSets){
# Split the gene sets into individual genes
split_genes <- function(gene_set) {
 unlist(strsplit(gene_set, split = "[,]"))
}

# Apply the splitting function to all significant gene sets
all_genes <- unique(unlist(lapply(sig.GeneSets, split_genes)))

# Annotate probes
top <- rownames_to_column(top, var = "ID")
anno <- as.data.frame(anno)
anno <- rownames_to_column(anno, var = "ID")
top <- merge(top, anno, by = "ID")

# Split UCSC_RefGene_Name into individual genes
top <- top %>%
  mutate(UCSC_RefGene_Name = strsplit(as.character(UCSC_RefGene_Name), split = "[,;]")) %>%
  unnest(UCSC_RefGene_Name)

# Filter for significant gene sets
filteredResults <- top[top$UCSC_RefGene_Name %in% all_genes, ]

# Summarize methylation status for each significant gene
methylationSummary <- sapply(all_genes, function(gene) {
 probes <- filteredResults[filteredResults$UCSC_RefGene_Name == gene, ]
 avgLogFC <- mean(probes$logFC)
 return(avgLogFC)
})

# Determine methylation status
methylationStatus <- ifelse(methylationSummary > 0, "Hypermethylated", "Hypomethylated")

# Combine results into a data frame
results <- data.frame(Gene = all_genes, MethylationStatus = methylationStatus, MethylationSummary=methylationSummary)

return(results)
}
```

```{r}
endmt.results <- sig.gene.list(sig.genes.endmt)
results <- sig.gene.list(sig.Genes.all)
```

## Pathway analysis of genes by methylation direction

```{r, echo = F}
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/Prakash Lab/TS Batch3 DNA Methylation Analysis/gene.list.R")
# create gene list of hypermethylated genes
hyper.gene.list <- gene.list("Hypermethylated")

# create gene list of hypomethylated genes
hypo.gene.list <- gene.list("Hypomethylated")
```

### GO

```{r}
if(nrow(top.go)>0){
# hypermethylated gene sets
go.hyper.enrich <- enrichGO(gene = hyper.gene.list, OrgDb = org.Hs.eg.db, keyType = "ENTREZID", ont = "ALL")

# hypomethylated gene sets
go.hypo.enrich <- enrichGO(gene = hypo.gene.list, OrgDb = org.Hs.eg.db, keyType = "ENTREZID", ont = "ALL")

# add column specifying methylation direction
go.hyper.enrich <- as.data.frame(go.hyper.enrich)
if (nrow(go.hyper.enrich)>0) {
  go.hyper.enrich$MethylationStatus = "Hypermethylated"
}

go.hypo.enrich <- as.data.frame(go.hypo.enrich)
if (nrow(go.hypo.enrich)>0) {
  go.hypo.enrich$MethylationStatus = "Hypomethylated"
}

# merge data sets
go.enrich <- rbind(go.hyper.enrich, go.hypo.enrich)
go.enrich <- go.enrich[rownames(go.enrich) %in% rownames(top.go),]
}
```

### KEGG

```{r}
if(nrow(top.kegg)>0){
# hypermethylated gene sets
kegg.hyper.enrich <- enrichKEGG(gene = hyper.gene.list, organism = "hsa")

# hypomethylated gene sets
kegg.hypo.enrich <- enrichKEGG(gene = hypo.gene.list, organism = "hsa")

# add column specifying methylation direction
kegg.hyper.enrich <- as.data.frame(kegg.hyper.enrich)
if (nrow(kegg.hyper.enrich)>0) {
  kegg.hyper.enrich$MethylationStatus = "Hypermethylated"
}

kegg.hypo.enrich <- as.data.frame(kegg.hypo.enrich)
if (nrow(kegg.hypo.enrich)>0) {
  kegg.hypo.enrich$MethylationStatus = "Hypomethylated"
} 

# merge data sets
kegg.enrich <- rbind(kegg.hyper.enrich, kegg.hypo.enrich)
kegg.enrich <- kegg.enrich[rownames(kegg.enrich) %in% rownames(top.kegg),]
}
```

# Visualize gene lists
```{r}
# pull cpgs associated with significant genes
anno.df <- as.data.frame(anno)
anno.df <- rownames_to_column(anno.df, var = "ID")

annotate_probes <- function(gene){
  top <- topTable(fit2, number = "Inf", coef = 2) %>% 
    rownames_to_column(var = "ID")
  top <- merge(top, anno.df, by = "ID") %>% # split gene list into individual genes
    dplyr::mutate(UCSC_RefGene_Name = strsplit(as.character(UCSC_RefGene_Name), split = "[,;]")) %>%
    unnest(UCSC_RefGene_Name)
  top1 <- top[top$UCSC_RefGene_Name==gene,]
  return(top1)
}

endmt.probes <- endmt.results %>%
  mutate(result = lapply(Gene, annotate_probes))
```

```{r}
# get beta values for each gene

sample_sheet <- sample_sheet %>% 
  mutate(Sentrix_Name = rownames(sample_sheet))

avg_Beta <- numeric(0)

for (gene in endmt.probes$Gene){
  
  cpgs_on_gene <- endmt.probes[endmt.probes$Gene==gene,]$result[[1]]$ID

  gene_beta <- beta_values[cpgs_on_gene,]

  # Convert matrix to long format
  gene_long <- as.data.frame(gene_beta) %>%
    rownames_to_column("probe_id") %>%
    pivot_longer(-probe_id, names_to = "Sentrix_Name", values_to = "beta") %>%
    left_join(sample_sheet, by = "Sentrix_Name") # merge subject info with beta values

  # Compute average methylation level of gene for each subject and average for each group
  group_means <- aggregate(beta ~ Karyotype, data = gene_long, FUN = mean)
  mean_diff <- group_means$beta[1] - group_means$beta[2]
  
  avg_Beta <- c(avg_Beta, mean_diff)
}

endmt.results$avg_Beta <- avg_Beta
```

```{r}
endmt.results <- endmt.results %>% 
  dplyr::filter(!is.na(MethylationStatus))
results <- results %>% 
  dplyr::filter(!is.na(MethylationStatus))

gene_plot <- function(res, descriptor){
  res <- res[1:50,]
gene_plot <- ggplot(res, aes(x = MethylationStatus, y = Gene)) +
  geom_point(aes(size = abs(avg_Beta), 
                 color = MethylationStatus), 
             alpha = 0.7) +
  scale_size_continuous(name = "Methylation\nLevel", guide = "none") +
  scale_color_manual(values = c("Hypermethylated" = "red", 
                               "Hypomethylated" = "blue"),
                     name = "MethylationStatus",
                     guide = "none") +
  theme_minimal(base_family = "serif") +
  labs(x = "Average Difference (Beta-value)", 
       y = "Gene",
       title = glue("Differentially Methylated Genes of {cell_phase}")) +
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)
ggsave(glue("gene_plot_{cell_phase}_{descriptor}.pdf"), gene_plot, width = 8, height = 8)
}

gene_plot(endmt.results, "ENDMT")
gene_plot(results, "ALL")
```

# Export Results

```{r}
if(exists("kegg.enrich")==TRUE){
  write.csv(kegg.enrich, glue("kegg_{cell_phase}_enrichment_results.csv"))
}

if(exists("go.enrich")==TRUE){
  write.csv(go.enrich, glue("go_{cell_phase}_enrichment_results.csv"))
  top.go <- topGSA(gst.region.go, number = "Inf")
}

write.csv(top.kegg, glue("{cell_phase}_kegg_region.csv"))
write.csv(top.go, glue("{cell_phase}_go_region.csv"))

# create csv of gene list
write.csv(results, glue("{cell_phase}_genes.csv"))
```
