---
title: "Longitudinal Analysis"
author: "Kennedy Zinn"
date: "2025-07-21"
output: html_document
---

```{r}
packages <- c("minfi", "limma", "DMRcate", "maxprobes", "readxl", "dplyr", "missMethyl", "MethylToSNP", "glue", "IlluminaHumanMethylationEPICanno.ilm10b5.hg38", "msigdbr", "tibble", "tidyr", "clusterProfiler", "org.Hs.eg.db", "purrr", "GenomicRanges", "ggplot2")
sapply(packages, library, character.only = TRUE)
```

```{r}
cell_phase = "iPSC" # either iPSC or EC
```

# Load data
```{r}
# load sample sheet
sample_sheet <- read_excel("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/TS/dna methyl/sample_sheet_ALL.xlsx",
                         skip = 9,
                         na="NA"
                         )
sample_sheet <- sample_sheet[,1:10] %>% 
  dplyr::filter(Phase==cell_phase) %>% 
  dplyr::filter(Karyotype!="46XisoXq")
rownames(sample_sheet) <- paste(sample_sheet$Sentrix_ID, sample_sheet$Sentrix_Position, sep = "_")
```

```{r}
# load the idat files
path_to_idat <- "/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/TS/dna methyl/SPMTS2_Array data/"

rgSet <- read.metharray.exp(path_to_idat)
```

# Pre-processing

```{r}
source("process_methyl_microarray.R")
gmSet <- process_methyl_microarray(rgSet)
```

Beta values, which range between [0,1], represent the proportion of methylated DNA copies at a specific CpG site compared to all of the DNA copies at that site. Beta values for the ith CpG site are mathematically defined as,

Beta(i) = max(yi(methyl),0) / [max(yi(unmethyl), 0) + max(yi(methyl), 0) + alpha],

where yi methyl and unmethyl are the intensities measured by the ith methylated and unmethylated probes, respectively. 0 ensures that negative signals that take place after background adjustment are not considered. Alpha, a constant (usually 100), serves as an offset to regularize beta when both methylated and unmethylated instensities are low. Beta values are intuitive, and hence useful for visual presentations of the reads. (Du et al., 2010)

```{r, echo = FALSE}
# Compute beta values
beta_values <- getBeta(gmSet) 
beta_values <- beta_values[,colnames(beta_values) %in% rownames(sample_sheet)]
```

M values are log transformed beta values. Negative M values indicate higher levels of unmethylated DNA copies than methylated DNA copies. Postive M values indicate higher levels of methylated DNA copies than unmethylated. M values close to zero indicate similar intensities between methylated and unmethylated probes. M values are mathematically expressed as,

M(i) = log2([max(yi(methyl))+alpha]/[max(yi(unmethyl))+alpha])

where alpha is a constant to offset dramatic changes resulting from small estimation errors. M values can be expressed in terms of beta values,

M(i) = log2(Beta(i)/[1 - Beta(i)])

and therefore, beta values can be expressed in terms of M values,

Beta(i) = 2\^M(i) / [2\^M(i) + 1]

M values are more statistically sound than beta values.

```{r}
# Compute M values
Mval <- getM(gmSet)
Mval <- Mval[,colnames(Mval) %in% rownames(sample_sheet)]
dim(Mval)
```


```{r}
# PCA of M values
pdf(glue("MDS_plot_{cell_phase}.pdf"))
par(mfrow=c(1,1), cex =.5)
plotMDS(Mval, labels=sample_sheet$Sample_Name, col=as.numeric(factor(sample_sheet$Karyotype, levels = c("46XX", "45X"))))
legend("top",legend=c("46XX", "45X"), col =1:2)
dev.off()
```

# Differential Methylation Analysis

```{r}
# match column names of Mval to row names of sample sheet
sample_sheet <- as.data.frame(sample_sheet)
rownames(sample_sheet) <- paste(sample_sheet$`Sentrix_ID`, sample_sheet$`Sentrix_Position`, sep = "_")

# reorder sample sheet to match row names to column names of Mval
sample_sheet <- sample_sheet[colnames(Mval),]
```

```{r}
#factor covariates
patient <- factor(sample_sheet$Patient)
disease <- factor(sample_sheet$Karyotype, levels = c("46XX", "45X"))

design <- model.matrix(~disease+patient)
design
```

```{r}
fit <- lmFit(Mval, design)
fit2 <- eBayes(fit)
top <- topTable(fit2, number = "Inf", coef="disease45X") # diff in karyotype, while controlling for patient
top
```

```{r}
# print proportion of sig CpGs hyper and hypo methylated
sig.cpg <- top[top$adj.P.Val<0.05,]
t <-table(sig.cpg$logFC<0)

hyper.proportion <- t[1]/nrow(sig.cpg)
hypo.proportion <- t[2]/nrow(sig.cpg)

glue("{nrow(sig.cpg)} probes are significantly differentially methylated in 45XO cells compared to 46XX cells. The proportion of hypermethylated CpG sites is {hyper.proportion}. The proportion of hypomethylated CpG sites is {hypo.proportion}.")
```

## Region Analysis
```{r}
myAnnotation <- cpg.annotate(datatype = "array", object = Mval, what = "M",
                             arraytype = "EPICv1",
                             analysis.type = "differential", design = design, contrasts = F, coef = 2)
DMRs <- suppressMessages(dmrcate(myAnnotation))
results.ranges <- extractRanges(DMRs, genome="hg38")
```

```{r}
anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)

gst.region.kegg <- goregion(results.ranges, all.cpg = rownames(top), collection = "KEGG", array.type="EPIC", anno=anno, sig.genes=T)
gst.region.go <- goregion(results.ranges, all.cpg = rownames(top), collection = "GO", array.type="EPIC", anno=anno, sig.genes=T)

top.go <- topGSA(gst.region.go, number = "Inf")
top.go <- top.go[top.go$FDR<0.05,]

top.kegg <- topGSA(gst.region.kegg, number = "Inf")
top.kegg <- top.kegg[top.kegg$FDR<0.05,]
```

```{r}
# Get the Hallmark gene sets for Homo sapiens (Human)
m_df <- msigdbr(species = "Homo sapiens", category = "H")

# Filter for the specific gene set and get the gene list
emt_gene_list <- m_df %>%
  filter(gs_name == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>%
  pull(entrez_gene)

# Put the gene list into a named list, as required by gsaregion
hallmark_gs_list <- list("HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION" = emt_gene_list)
```

```{r}
gsa.region <- gsaregion(results.ranges, all.cpg = rownames(top), collection = hallmark_gs_list, array.type="EPIC", anno=anno, sig.genes=T)
```

## Identify the direction of significant genes
```{r}
sig.Genes.all <- unique(c(gst.region.go$SigGenesInSet,
                         gst.region.kegg$SigGenesInSet))
sig.genes.endmt <- unique(gsa.region$SigGenesInSet)

sig.gene.list <- function(sig.GeneSets){
# Split the gene sets into individual genes
split_genes <- function(gene_set) {
 unlist(strsplit(gene_set, split = "[,]"))
}

# Apply the splitting function to all significant gene sets
all_genes <- unique(unlist(lapply(sig.GeneSets, split_genes)))

# Annotate probes
top <- rownames_to_column(top, var = "ID")
anno <- as.data.frame(anno)
anno <- rownames_to_column(anno, var = "ID")
top <- merge(top, anno, by = "ID")

# Split UCSC_RefGene_Name into individual genes
top <- top %>%
  mutate(UCSC_RefGene_Name = strsplit(as.character(UCSC_RefGene_Name), split = "[,;]")) %>%
  unnest(UCSC_RefGene_Name)

# Filter for significant gene sets
filteredResults <- top[top$UCSC_RefGene_Name %in% all_genes, ]

# get beta values for each group
sample_45 <- sample_sheet %>% 
  dplyr::filter(Karyotype=="45X")
sample_46 <- sample_sheet %>% 
  dplyr::filter(Karyotype=="46XX")

beta_45 <- beta_values[, colnames(beta_values) %in% rownames(sample_45)]
beta_46 <- beta_values[, colnames(beta_values) %in% rownames(sample_46)]

# Summarize methylation status and beta value for each significant gene
methylationSummary <- sapply(all_genes, function(gene) {
 probes <- filteredResults[filteredResults$UCSC_RefGene_Name == gene, ]
 
 avgLogFC <- mean(probes$logFC)
 
 beta_45_g <- beta_45[rownames(beta_45) %in% probes$ID,]
 row_mean_beta_45 <- rowMeans(beta_45_g)
 avgBeta_45 <- mean(row_mean_beta_45)
 
 beta_46_g <- beta_46[rownames(beta_46) %in% probes$ID,]
 row_mean_beta_46 <- rowMeans(beta_46_g)
 avgBeta_46 <- mean(row_mean_beta_46)
 
 avgBeta_diff <- avgBeta_45 - avgBeta_46
 return(c(avgBeta_diff = avgBeta_diff, avg_M_diff=avgLogFC))
})

methylationSummary <- t(methylationSummary)

# Determine methylation status
methylationStatus <- ifelse(methylationSummary[,"avgBeta_diff"] > 0, "Hypermethylated", "Hypomethylated")

# Combine results into a data frame
results <- data.frame(Gene = all_genes, MethylationStatus = methylationStatus, avg_M_diff=methylationSummary[, "avg_M_diff"], avg_Beta_diff = methylationSummary[, "avgBeta_diff"])

return(results)
}
```

```{r}
endmt.results <- sig.gene.list(sig.genes.endmt)
results <- sig.gene.list(sig.Genes.all)
```

# Visualize gene lists

```{r}
endmt.results <- endmt.results %>% 
  dplyr::filter(!is.na(MethylationStatus))
results <- results %>% 
  dplyr::filter(!is.na(MethylationStatus))

gene_plot <- function(res, descriptor){
gene_plot <- ggplot(res, aes(x = MethylationStatus, y = Gene)) +
  geom_point(aes(size = abs(avg_Beta_diff), 
                 color = MethylationStatus), 
             alpha = 0.7) +
  scale_size_continuous(name = "Methylation\nLevel", guide = "none") +
  scale_color_manual(values = c("Hypermethylated" = "red", 
                               "Hypomethylated" = "blue"),
                     name = "MethylationStatus",
                     guide = "none") +
  theme_minimal(base_family = "serif") +
  labs(x = "Average Difference (Beta-value)", 
       y = "Gene",
       title = glue("Differentially Methylated Genes of {cell_phase}")) +
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)
ggsave(glue("gene_plot_{cell_phase}_{descriptor}.pdf"), gene_plot, width = 8, height = 8)
}

gene_plot(endmt.results, "ENDMT")
gene_plot(results, "ALL")

#get top 50 genes
results <- results %>% 
  dplyr::arrange(desc(avg_Beta_diff))

res_50 <- results[1:50,]
gene_plot(res_50, "top50")
```

# Export Results

```{r}
write.csv(top.kegg, glue("{cell_phase}_kegg_region.csv"))
write.csv(top.go, glue("{cell_phase}_go_region.csv"))

# create csv of gene list
write.csv(results, glue("{cell_phase}_genes.csv"))
```
